### ISSUE TITLE
ERROR - lazrs failed to decompress points: lazrs is not installed

### ISSUE THREAD

**Issue by skoulouzis** [QCDIS/NaaVRE#149]
Laserfarm notebbok cannot normalize all the point cloud 
ERROR - lazrs failed to decompress points: lazrs is not installed
```python
# Configurations 
import fnmatch
import json
import getpass
import os
import pathlib
import datetime
                    
from dask.distributed import LocalCluster, SSHCluster 
from laserfarm import Retiler, DataProcessing, GeotiffWriter, MacroPipeline
from laserfarm.remote_utils import get_wdclient, get_info_remote, list_remote

conf_remote_path_root = pathlib.Path('/webdav')
conf_remote_path_ahn = pathlib.Path('/webdav/ahn')
conf_remote_path_retiled = pathlib.Path('/webdav/retiled/')
conf_remote_path_norm = pathlib.Path('/webdav/norm/')
conf_remote_path_targets = pathlib.Path('/webdav/targets')
conf_local_tmp = pathlib.Path('/tmp')

param_hostname = ''
param_login = ''
param_password = ''

param_feature_name = 'perc_95_normalized_height'
param_validate_precision = '0.00001'
param_tile_mesh_size = '10.'
param_filter_type= 'select_equal'
param_attribute = 'raw_classification'
param_min_x = '-113107.81'
param_max_x = '398892.19'
param_min_y = '214783.87'
param_max_y = '726783.87'
param_n_tiles_side = '512'
param_apply_filter_value = '1'
param_laz_compression_factor = '7'

conf_wd_opts = { 'webdav_hostname': param_hostname, 'webdav_login': param_login, 'webdav_password': param_password}

laz_files = [f for f in list_remote(get_wdclient(conf_wd_opts), conf_remote_path_ahn.as_posix())
             if f.lower().endswith('.laz')]

max_filesize = 250 * 2**20  # desired max file size (in bytes)
def save_chunk_to_laz_file(in_filename, 
                           out_filename, 
                           offset, 
                           n_points):
    """Read points from a LAS/LAZ file and write them to a new file."""
    
    points = np.array([])
    
    with laspy.open(in_filename) as in_file:
        with laspy.open(out_filename, 
                        mode="w", 
                        header=in_file.header) as out_file:
            in_file.seek(offset)
            points = in_file.read_points(n_points)
            out_file.write_points(points)
    return len(points)

def split_strategy(filename, max_filesize):
    """Set up splitting strategy for a LAS/LAZ file."""
    with laspy.open(filename) as f:
        bytes_per_point = (
            f.header.point_format.num_standard_bytes +
            f.header.point_format.num_extra_bytes
        )
        n_points = f.header.point_count
    n_points_target = int(
        max_filesize * int(param_laz_compression_factor) / bytes_per_point
    )
    stem, ext = os.path.splitext(filename)
    return [
        (filename, f"{stem}-{n}{ext}", offset, n_points_target)
        for n, offset in enumerate(range(0, n_points, n_points_target))
    ]

from IPython.display import display, HTML   

remote_path_retiled = str(conf_remote_path_retiled)

grid_retile = {
    'min_x': float(param_min_x),
    'max_x': float(param_max_x),
    'min_y': float(param_min_y),
    'max_y': float(param_max_y),
    'n_tiles_side': int(param_n_tiles_side)
}


retiling_input = {
    'setup_local_fs': {'tmp_folder': conf_local_tmp.as_posix()},
    'pullremote': conf_remote_path_ahn.as_posix(),
    'set_grid': grid_retile,
    'split_and_redistribute': {},
    'validate': {},
    'pushremote': conf_remote_path_retiled.as_posix(),
    'cleanlocalfs': {}
}
    
file = laz_files[0]
    
retiler = Retiler(file,label=file).config(retiling_input).setup_webdav_client(conf_wd_opts)
retiler_output = retiler.run()

remote_path_retiled

tiles = [t.strip('/') for t in list_remote(get_wdclient(conf_wd_opts), conf_remote_path_retiled.as_posix())
         if fnmatch.fnmatch(t, 'tile_*_*/')]

import copy

tiles

path_output = conf_remote_path_norm
normalization_input = {
    'setup_local_fs': {'tmp_folder': conf_local_tmp.as_posix()},
    'pullremote': conf_remote_path_retiled.as_posix(),
    'load': {'attributes': 'all'},
    # Filter out artifically high points - give overflow error when writing
    'apply_filter': {'filter_type':'select_below',
                     'attribute': 'z',
                     'threshold': 10000.},  # remove non-physically heigh points
    'normalize': 1,
    'clear_cache' : {},
    'pushremote': conf_remote_path_norm.as_posix(),
}

# write input dictionary to JSON file
with open('normalize.json', 'w') as f:
    json.dump(normalization_input, f)
    

# add pipeline list to macro-pipeline object and set the corresponding labels
tile = tiles[0]
Retiler(file,label=file).config(retiling_input).setup_webdav_client(conf_wd_opts)
normalization_input_ = copy.deepcopy(normalization_input)
normalization_input_['export_point_cloud'] = {'filename': '{}.laz'.format(tile),
                                              'overwrite': True}
dp = DataProcessing(tile, label=tile).config(normalization_input_).setup_webdav_client(conf_wd_opts)
dp.run()
```

**Comment by skoulouzis** [comment]
when installing  lazrs==0.4.0
```python
2022-05-17 20:10:38,556 -           laserfarm.pipeline_remote_data -       INFO - Input dir set to /tmp/tile_278_391_input
2022-05-17 20:10:38,558 -           laserfarm.pipeline_remote_data -       INFO - Output dir set to /tmp/tile_278_391_output
2022-05-17 20:10:38,560 -           laserfarm.pipeline_remote_data -       INFO - Pulling from WebDAV /webdav/retiled/tile_278_391 ...
2022-05-17 20:10:41,370 -           laserfarm.pipeline_remote_data -       INFO - ... pulling completed.
2022-05-17 20:10:41,371 -                laserfarm.data_processing -       INFO - Loading point cloud data ...
2022-05-17 20:10:41,372 -                laserfarm.data_processing -       INFO - ... loading /tmp/tile_278_391_input/tile_278_391/C_01GN2_1.LAZ
thread '<unnamed>' panicked at 'assertion failed: mid <= self.len()', /rustc/514b3877956dc594823106b66c164f8cdbc8b3da/library/core/src/slice/mod.rs:1537:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
Input In [14], in <cell line: 32>()
     29 normalization_input_['export_point_cloud'] = {'filename': '{}.laz'.format(tile),
     30                                               'overwrite': True}
     31 dp = DataProcessing(tile, label=tile).config(normalization_input_).setup_webdav_client(conf_wd_opts)
---> 32 dp.run()

File /opt/conda/lib/python3.9/site-packages/laserfarm/pipeline_remote_data.py:113, in PipelineRemoteData.run(self, pipeline)
    109 _pipeline = pipeline if pipeline is not None else self.pipeline
    110 _pipeline = (('setup_local_fs', 'setup_webdav_client', 'pullremote')
    111              + _pipeline
    112              + ('pushremote', 'cleanlocalfs'))
--> 113 super(PipelineRemoteData, self).run(pipeline=_pipeline)

File /opt/conda/lib/python3.9/site-packages/laserfarm/pipeline.py:108, in Pipeline.run(self, pipeline)
    106 input_task = _input.pop(task_name)
    107 if isinstance(input_task, dict):
--> 108     task(**input_task)
    109 elif (isinstance(input_task, list)
    110       or isinstance(input_task, tuple)):
    111     task(*input_task)

File /opt/conda/lib/python3.9/site-packages/laserfarm/data_processing.py:104, in DataProcessing.load(self, **load_opts)
    102 for file in input_file_list:
    103     logger.info('... loading {}'.format(file))
--> 104     add_to_point_cloud(self.point_cloud, load(file, **load_opts))
    105 logger.info('... loading completed.')
    106 return self

File /opt/conda/lib/python3.9/site-packages/laserchicken/io/load.py:17, in load(path, format, *args, **kwargs)
      7 """
      8 Read point cloud from a file.
      9 
   (...)
     14 :return: point cloud data
     15 """
     16 reader = get_io_handler(path, mode='r', format=format)
---> 17 point_cloud = reader.read(*args, **kwargs)
     18 add_metadata(point_cloud, sys.modules[__name__], {'path': path, 'args': args, **kwargs})
     19 return point_cloud

File /opt/conda/lib/python3.9/site-packages/laserchicken/io/las_handler.py:29, in LASHandler.read(self, attributes)
     22 def read(self, attributes=DEFAULT_LAS_ATTRIBUTES):
     23     """
     24     Load the points from a LAS(LAZ) file into memory.
     25 
     26     :param attributes: list of attributes to read ('all' for all attributes in file)
     27     :return: point cloud data structure
     28     """
---> 29     file = pylas.read(self.path)
     30     attributes_available = [el if el not in ['X', 'Y', 'Z'] else el.lower()
     31                             for el in file.points.dtype.names]
     33     attributes = select_valid_attributes(attributes_available, attributes)

File /opt/conda/lib/python3.9/site-packages/pylas/lib.py:94, in read_las(source, closefd)
     70 """ Entry point for reading las data in pylas
     71 
     72 Reads the whole file into memory.
   (...)
     91     The object you can interact with to get access to the LAS points & VLRs
     92 """
     93 with open_las(source, closefd=closefd) as reader:
---> 94     return reader.read()

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:64, in LasReader.read(self)
     61 self.stream.seek(self.start_pos + self.header.offset_to_point_data)
     63 try:
---> 64     points = self._read_points(vlrs)
     65 except errors.LazError as e:
     66     logger.error("error when decompressing {}, trying laszip".format(e))

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:114, in LasReader._read_points(self, vlrs)
    112 if self.header.are_points_compressed:
    113     laszip_vlr = vlrs.pop(vlrs.index("LasZipVlr"))
--> 114     points = self._read_compressed_points_data(laszip_vlr, point_format)
    115 else:
    116     points = record.PackedPointRecord.from_stream(
    117         self.stream, point_format, self.header.point_count
    118     )

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:136, in LasReader._read_compressed_points_data(self, laszip_vlr, point_format)
    133 struct.pack_into("<q", points_data, 0, offset_to_chunk_table)
    135 try:
--> 136     decompressed_points = lazrs_decompress_buffer(
    137         points_data,
    138         point_format.dtype.itemsize,
    139         self.header.point_count,
    140         laszip_vlr
    141     )
    142 except errors.LazError as e:
    143     logger.error("lazrs failed to decompress points: {}".format(e))

File /opt/conda/lib/python3.9/site-packages/pylas/compression.py:64, in lazrs_decompress_buffer(compressed_buffer, point_size, point_count, laszip_vlr, parallel)
     60     vlr_data = np.frombuffer(laszip_vlr.record_data, dtype=np.uint8)
     62     point_decompressed = np.zeros(point_count * point_size, np.uint8)
---> 64     lazrs.decompress_points(point_compressed, vlr_data, point_decompressed, parallel)
     65 except lazrs.LazrsError as e:
     66     raise LazError("lazrs error: {}".format(e)) from e

PanicException: assertion failed: mid <= self.len()
```

**Comment by skoulouzis** [comment]
when installing lazrs==0.2.1 or lazrs==0.3.0
```python
2022-05-17 20:10:38,556 -           laserfarm.pipeline_remote_data -       INFO - Input dir set to /tmp/tile_278_391_input
2022-05-17 20:10:38,558 -           laserfarm.pipeline_remote_data -       INFO - Output dir set to /tmp/tile_278_391_output
2022-05-17 20:10:38,560 -           laserfarm.pipeline_remote_data -       INFO - Pulling from WebDAV /webdav/retiled/tile_278_391 ...
2022-05-17 20:10:41,370 -           laserfarm.pipeline_remote_data -       INFO - ... pulling completed.
2022-05-17 20:10:41,371 -                laserfarm.data_processing -       INFO - Loading point cloud data ...
2022-05-17 20:10:41,372 -                laserfarm.data_processing -       INFO - ... loading /tmp/tile_278_391_input/tile_278_391/C_01GN2_1.LAZ
thread '<unnamed>' panicked at 'assertion failed: mid <= self.len()', /rustc/514b3877956dc594823106b66c164f8cdbc8b3da/library/core/src/slice/mod.rs:1537:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
Input In [14], in <cell line: 32>()
     29 normalization_input_['export_point_cloud'] = {'filename': '{}.laz'.format(tile),
     30                                               'overwrite': True}
     31 dp = DataProcessing(tile, label=tile).config(normalization_input_).setup_webdav_client(conf_wd_opts)
---> 32 dp.run()

File /opt/conda/lib/python3.9/site-packages/laserfarm/pipeline_remote_data.py:113, in PipelineRemoteData.run(self, pipeline)
    109 _pipeline = pipeline if pipeline is not None else self.pipeline
    110 _pipeline = (('setup_local_fs', 'setup_webdav_client', 'pullremote')
    111              + _pipeline
    112              + ('pushremote', 'cleanlocalfs'))
--> 113 super(PipelineRemoteData, self).run(pipeline=_pipeline)

File /opt/conda/lib/python3.9/site-packages/laserfarm/pipeline.py:108, in Pipeline.run(self, pipeline)
    106 input_task = _input.pop(task_name)
    107 if isinstance(input_task, dict):
--> 108     task(**input_task)
    109 elif (isinstance(input_task, list)
    110       or isinstance(input_task, tuple)):
    111     task(*input_task)

File /opt/conda/lib/python3.9/site-packages/laserfarm/data_processing.py:104, in DataProcessing.load(self, **load_opts)
    102 for file in input_file_list:
    103     logger.info('... loading {}'.format(file))
--> 104     add_to_point_cloud(self.point_cloud, load(file, **load_opts))
    105 logger.info('... loading completed.')
    106 return self

File /opt/conda/lib/python3.9/site-packages/laserchicken/io/load.py:17, in load(path, format, *args, **kwargs)
      7 """
      8 Read point cloud from a file.
      9 
   (...)
     14 :return: point cloud data
     15 """
     16 reader = get_io_handler(path, mode='r', format=format)
---> 17 point_cloud = reader.read(*args, **kwargs)
     18 add_metadata(point_cloud, sys.modules[__name__], {'path': path, 'args': args, **kwargs})
     19 return point_cloud

File /opt/conda/lib/python3.9/site-packages/laserchicken/io/las_handler.py:29, in LASHandler.read(self, attributes)
     22 def read(self, attributes=DEFAULT_LAS_ATTRIBUTES):
     23     """
     24     Load the points from a LAS(LAZ) file into memory.
     25 
     26     :param attributes: list of attributes to read ('all' for all attributes in file)
     27     :return: point cloud data structure
     28     """
---> 29     file = pylas.read(self.path)
     30     attributes_available = [el if el not in ['X', 'Y', 'Z'] else el.lower()
     31                             for el in file.points.dtype.names]
     33     attributes = select_valid_attributes(attributes_available, attributes)

File /opt/conda/lib/python3.9/site-packages/pylas/lib.py:94, in read_las(source, closefd)
     70 """ Entry point for reading las data in pylas
     71 
     72 Reads the whole file into memory.
   (...)
     91     The object you can interact with to get access to the LAS points & VLRs
     92 """
     93 with open_las(source, closefd=closefd) as reader:
---> 94     return reader.read()

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:64, in LasReader.read(self)
     61 self.stream.seek(self.start_pos + self.header.offset_to_point_data)
     63 try:
---> 64     points = self._read_points(vlrs)
     65 except errors.LazError as e:
     66     logger.error("error when decompressing {}, trying laszip".format(e))

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:114, in LasReader._read_points(self, vlrs)
    112 if self.header.are_points_compressed:
    113     laszip_vlr = vlrs.pop(vlrs.index("LasZipVlr"))
--> 114     points = self._read_compressed_points_data(laszip_vlr, point_format)
    115 else:
    116     points = record.PackedPointRecord.from_stream(
    117         self.stream, point_format, self.header.point_count
    118     )

File /opt/conda/lib/python3.9/site-packages/pylas/lasreader.py:136, in LasReader._read_compressed_points_data(self, laszip_vlr, point_format)
    133 struct.pack_into("<q", points_data, 0, offset_to_chunk_table)
    135 try:
--> 136     decompressed_points = lazrs_decompress_buffer(
    137         points_data,
    138         point_format.dtype.itemsize,
    139         self.header.point_count,
    140         laszip_vlr
    141     )
    142 except errors.LazError as e:
    143     logger.error("lazrs failed to decompress points: {}".format(e))

File /opt/conda/lib/python3.9/site-packages/pylas/compression.py:64, in lazrs_decompress_buffer(compressed_buffer, point_size, point_count, laszip_vlr, parallel)
     60     vlr_data = np.frombuffer(laszip_vlr.record_data, dtype=np.uint8)
     62     point_decompressed = np.zeros(point_count * point_size, np.uint8)
---> 64     lazrs.decompress_points(point_compressed, vlr_data, point_decompressed, parallel)
     65 except lazrs.LazrsError as e:
     66     raise LazError("lazrs error: {}".format(e)) from e

PanicException: assertion failed: mid <= self.len()
```

**Comment by fnattino** [comment]
@skoulouzis I think this an incompatibility between the backends of laspy (used in the splitting) and plyas (employed by laserchicken, now deprecated in favour of laspy). I'll update laserchicken to use laspy as well and make a new release!

**Comment by fnattino** [comment]
Hi @skoulouzis, I have just released a new version of laserchicken (0.5.0) with the fix that should solve this problem

**Comment by skoulouzis** [comment]
@fnattino  Yes I checked and it is now fixed. Thanks.
This is the output:
```
2022-05-22 11:44:04,565 -           laserfarm.pipeline_remote_data -       INFO - Input dir set to /tmp/tile_278_391_input
2022-05-22 11:44:04,565 -           laserfarm.pipeline_remote_data -       INFO - Output dir set to /tmp/tile_278_391_output
2022-05-22 11:44:04,570 -           laserfarm.pipeline_remote_data -       INFO - Pulling from WebDAV /webdav/retiled/tile_278_391 ...
2022-05-22 11:44:08,363 -           laserfarm.pipeline_remote_data -       INFO - ... pulling completed.
2022-05-22 11:44:08,366 -                laserfarm.data_processing -       INFO - Loading point cloud data ...
2022-05-22 11:44:08,368 -                laserfarm.data_processing -       INFO - ... loading /tmp/tile_278_391_input/tile_278_391/C_01GN2_1.LAZ
2022-05-22 11:44:09,931 -                laserfarm.data_processing -       INFO - ... loading completed.
2022-05-22 11:44:09,936 -                laserfarm.data_processing -       INFO - Normalizing point-cloud heights ...
2022-05-22 11:44:10,795 -                                     root -       INFO - Cylinder size in Bytes: 3056136558.191318
2022-05-22 11:44:10,799 -                                     root -       INFO - Memory size in Bytes: 7113195520
2022-05-22 11:44:10,801 -                                     root -       INFO - Start tree creation
2022-05-22 11:44:11,206 -                                     root -       INFO - Done with env tree creation
2022-05-22 11:44:11,262 -                                     root -       INFO - Done with target tree creation
2022-05-22 11:44:30,033 -                laserfarm.data_processing -       INFO - ... normalization completed.
2022-05-22 11:44:30,036 -                laserfarm.data_processing -       INFO - Filtering point-cloud data
2022-05-22 11:44:30,231 -                laserfarm.data_processing -       INFO - Exporting environment point-cloud ...
2022-05-22 11:44:30,235 -                laserfarm.data_processing -       INFO - ... exporting /tmp/tile_278_391_output/tile_278_391.laz
2022-05-22 11:44:32,121 -                laserfarm.data_processing -       INFO - ... exporting completed.
2022-05-22 11:44:32,124 -                laserfarm.data_processing -       INFO - Clearing cached KDTrees ...
2022-05-22 11:44:32,126 -           laserfarm.pipeline_remote_data -       INFO - Pushing to WebDAV /webdav/norm ...
2022-05-22 11:44:37,901 -           laserfarm.pipeline_remote_data -       INFO - ... pushing completed.
```
